ironfish/src/rpc/adapters/httpAdapter.ts

```diff
@@ -0,0 +1,184 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+import http from 'http'
+import { v4 as uuid } from 'uuid'
+import { Assert } from '../../assert'
+import { Logger } from '../../logger'
+import { ErrorUtils } from '../../utils'
+import { RpcRequest } from '../request'
+import { ApiNamespace, Router } from '../routes'
+import { RpcServer } from '../server'
+import { IRpcAdapter } from './adapter'
+import { ResponseError } from './errors'
+
+export class RpcHttpAdapter implements IRpcAdapter {
+  server: http.Server | null = null
+  router: Router | null = null
+
+  readonly host: string
+  readonly port: number
+  readonly logger: Logger
+  readonly namespaces: ApiNamespace[]
+  readonly maxRequestSize: number
+  private requests: Map<
+    string,
+    {
+      rpcRequest?: RpcRequest
+      req: http.IncomingMessage
+    }
+  >
+
+  constructor(host: string, port: number, logger: Logger, namespaces: ApiNamespace[]) {
+    this.host = host
+    this.port = port
+    this.logger = logger
+    this.namespaces = namespaces
+    this.requests = new Map()
+    // TODO(daniel): do we need a max message size and what should it be?
+    this.maxRequestSize = 5 * 1000 * 1000
+  }
+
+  attach(server: RpcServer): void | Promise<void> {
+    this.router = server.getRouter(this.namespaces)
+  }
+
+  start(): Promise<void> {
+    this.logger.debug(`Serving RPC on HTTP ${this.host}:${this.port}`)
+
+    const server = http.createServer()
+    this.server = server
+
+    return new Promise((resolve, reject) => {
+      const onError = (err: unknown) => {
+        server.off('error', onError)
+        server.off('listening', onListening)
+        reject(err)
+      }
+
+      const onListening = () => {
+        server.off('error', onError)
+        server.off('listening', onListening)
+
+        server.on('request', (req, res) => {
+          const reqId = uuid()
+          this.requests.set(reqId, { req })
+
+          req.on('close', () => {
+            this.cleanUpRequest(reqId)
+          })
+
+          void this.handleRequest(req, res, reqId).catch((e) => {
+            // TODO: handle (e instanceof ResponseError) here?
+            const error = ErrorUtils.renderError(e)
+            this.logger.debug(`Error in HTTP adapter: ${error}`)
+            res.writeHead(500)
+            res.end(JSON.stringify({ error }))
```

Comment:
This is a non RPC uniform error. We should return a uniform RPC error so they don't have to handle this specially.

Reply:
Yea good suggestion. modified to handle more of our error types